### PR TITLE
[WIP] Experimental use of openexr C layer in openimageio

### PR DESCRIPTION
**Pull request by kdt3rd**
Signed-off-by: Kimball Thurston <kdt3rd@gmail.com>

## Description

Experimental only version of use of the C layer for reading images, to be used for evaluating the new API, and testing improved thread behavior when higher level application (i.e. renderer) has it's own pile of threads hammering away at textures.

## Tests

At this point, just smoke test that iv can load a couple of images.

## Checklist:

- [x] I have read the [contribution guidelines](../CONTRIBUTING.md).
- [ ] If this is more extensive than a small change to existing code, I
  have previously submitted a Contributor License Agreement
  ([individual](../src/doc/CLA-INDIVIDUAL), and if there is any way my
  employers might think my programming belongs to them, then also
  [corporate](../src/doc/CLA-CORPORATE)).
- [x] I have updated the documentation, if applicable.
- [x] I have ensured that the change is tested somewhere in the testsuite
  (adding new test cases if necessary).
- [x] My code follows the prevailing code style of this project.

### CODE REVIEW

**Comment by kdt3rd** [comment]
lg - I believe I've re-done the locks that'll be used in exrinput_c.cpp to be more conducive to not locking during texturesys, except when the internal cache of the imagespec needs to be built (although we could in theory get rid of that and do the channel remap slightly differently)

File: src/openexr.imageio/exrinput_c.cpp
Comment (lgritz) [inline src/openexr.imageio/exrinput_c.cpp:538]: The diffs in this area make me wonder: Would it be a better choice if the C version uses structs in such a way that you say `top_datawindow.min.x` rather than `.x_min`, and then there's a lot more code that looks identical between the C and C++ versions? And you could even have handy functions (on the C++ side) that convert between the C and C++ equivalents of many of the imath classes.

**Review by lgritz — APPROVED** [review]
Looks good. I'm looking forward to trying this out.

File: src/openexr.imageio/exrinput_c.cpp
Comment (kdt3rd) [reply src/openexr.imageio/exrinput_c.cpp:451]: awesome, will clean that out then :)

File: src/openexr.imageio/exrinput_c.cpp
Comment (kdt3rd) [reply src/openexr.imageio/exrinput_c.cpp:461]: yeah, that comment was around when the user does not provide a proxy - I like to think about less code, less bugs as a first course of action. But I suppose if the implementor changes the default behaviour of IOProxy, this will continue to leave this behaving consistently. Given I can remove the exception bits, this will look much simpler after

File: src/openexr.imageio/exrinput_c.cpp
Comment (kdt3rd) [reply src/openexr.imageio/exrinput_c.cpp:538]: i debated that. You can certainly memory cast the structs, and I have unit tests to do so, and with your new constructors to imath, the vectors should just copy construct. I chose the "different" route to drive home that I'm not providing any of the other functionality in the c++ object, but maybe that is not the best way to indicate that

Comment (lgritz) [reply src/openexr.imageio/exrinput_c.cpp:538]: It's your code and your call. But I think I would have chosen to make the analogs of the C++ structures organized such that in client code, the API calls obviously would be different between the C and C++ interfaces, but as much as possible of the secondary, more pedestrian code were identical for C and C++ in all the places that they could be  (like how you access struct members). It's not something I'm going to lose sleep over, but because it would change source compatibility, it's pretty much now or never to lock these choices in or rethink them.

BTW, I added the interop utilities only for Vec and Matrix types, but maybe it would be handy for Box, too.

**Comment by lgritz** [comment]
Should we merge this, or do you have other changes coming right away?

Are you brave enough to edit `.github/workflows/ci.yml` to enable this code path in the "bleeding edge" test (which pulls from OpenEXR master) so that this is going through CI henceforth? We'd also need to exclude any testsuite entries that use deep images. If you don't want to mess with that, I can do it after the merge and submit a separate PR.

**Comment by kthurston** [comment]
I can add a CI path for this - wasn't sure whether you wanted to keep it as a separate file (i.e. two implementations)? I can put it as ifdef blocks in exrinput if that would be better? Just thought it was easier to compare side-by-side while we play. Did you have a look at what I did for the tile reads? I never made a multi-tile read call in oiio (only ever single tile at a time), so wasn't sure I interpreted the desired memory layout (interleaved channels, tiles are successive in memory in y, x order. Also - in the new world, if you are at a small mip level (i.e. smaller than the tile size, it only negotiates that tile size in memory inside the C code, and hadn't validated the spec is correct for that yet. - I have the accessors I made to query the sizes from the internal structs in exr, probably should use those to ensure that. Further, wasn't sure that is what the API is expected to be filling in, or if it's expecting to always deliver a full tilesize (i.e. 32x32) regardless of mip level

**Comment by lgritz** [comment]
> wasn't sure whether you wanted to keep it as a separate file (i.e. two implementations)? I can put it as ifdef blocks in exrinput if that would be better? Just thought it was easier to compare side-by-side while we play. 

I think this is great, and at this stage made it easy to diff the two files. Let's keep it separate until it's complete and as we fiddle around, but then eventually when it just does this by default, we'll want to merge this all into one file with appropriate `#if` guards.

> I never made a multi-tile read call in oiio (only ever single tile at a time), 

Putting it into CI will uncover any problems immediately, as there are plenty of tests involving tiled images and the multi-tile reads happen whenever we read whole images rather than single tiles in the texture system.

Also, in `iv`, the `<` and `>` will show you the different MIP levels of an image.

> But at least let's get the exception bits cleaned up - I started that this morning, but meetings are starting now...

Up to you. Nothing wrong with coming back to simplify that part later.

**Comment by lgritz** [comment]
@etheory We are getting close to being able to benchmark the new OpenEXR code for better threaded tile performance.

**Comment by kthurston** [comment]
ok, I'll push up the exception cleanup and the CI path tonight / tomorrow morning, then we can unleash the hounds as they say

**Comment by kdt3rd** [comment]
got things patched up, started running tests, there are a couple of problems, let me get those patched up prior to merging this in

**Comment by kdt3rd** [comment]
fyi, when I run the tests, I see a number of failures with tiff and with freetype, it seems to use a different font on my system than for you (ubuntu?) by default. However, there are a couple of bad / incomplete tests that fail because the error messages are different. I'm happy to update the references, but wanted you to have a look prior to doing so to make sure I am interpreting it correctly.

**Comment by kdt3rd** [comment]
additionally, I did visually audit the texture sys chain, so other than the lock that happens when you first access a subimage / miplevel to update the ImageSpec cache in the input file, I believe all the locking should be removed, so should be ready for some baseline tests. Will love to hear how it helps, or if there are other issues.

**Comment by lgritz** [comment]
Freetype has the maddening property of rasterizing fonts ever so slightly differently from version to version. This is one case where we tend to just check multiple reference images into the testsuite/NAME/ref directory (the testsuite scripts will allow a pass if the output matches *any* of the reference images in that directory). If your Linux distro has a different version of Freetype than any of the ones we've encountered in CI or our personal machines, it may indeed differ in output. Feel free (maybe in a separate PR?) to add new reference output for your version of freetype. You can see how we usually name the reference files to make it clear which freetype versions required new reference images.
```
$ ls testsuite/oiiotool-text/ref
aligned-freetype2.4.11.tif  text-freetype2.7.tif
aligned.tif                 text.tif
out.txt                     textalpha.tif
text-alt.tif                textshadowed-freetype2.4.11.tif
text-freetype2.6.5.tif      textshadowed.tif
```

**Comment by kdt3rd** [comment]
fyi, the image tests that OpenImageIO does has uncovered an issue with the piz decompressor under C beyond the current unit tests in the exr tree. Ignoring that, I believe I have fixed the issues I've found with the texture reading in OpenImageIO and the new C library.

I will fix the piz decompression tomorrow and wire in deep tomorrow. But I don't want to push up reference output changes from the different error text, so we'll continue to have errors even after I fix those.

side note: I found out why the text tests were failing under Arch: I didn't have the droid fonts installed, so it wasn't using DroidSans. So seems like the newer freetype (2.10.4) hasn't changed much. I don't seem to have the correct DroidSerif font for one of the oiiotool tests (also not found), but not going to worry too much about that right now.

**Comment by lgritz** [comment]
The Droid fonts are installed as part of OIIO itself. I suspect that if you only build (but don't install) and then run tests, maybe it can't find them. Try just setting OpenImageIO_ROOT to wherever you have your test install of OIIO, and that should be enough to find them. I'll try to make the testsuite itself a little more robust to not needing that set if you have a build area.

**Comment by kdt3rd** [comment]
@lgritz - There may still be issues, but this (with latest openexr master) seems close enough for now to enable wider testing

**Comment by kthurston** [comment]
Huh, I see that unit_imageinout is crashing, will take a look at that tonight. I believe the other two failing tests are differences in the error messages.

**Comment by kdt3rd** [comment]
good to go whenever I think - the C layer isn't designed to read non-chunk-sized blocks, but think I've fixed the various corner cases I've seen

**Comment by lgritz** [comment]
I have a "fix" for the error messages. By which I mean, a patch that adds an additional reference output for it to match to for these two tests. I'm testing it now and will push to your branch if it succeeds.

But in the mean time, I do urge you to look at the new error output reference versus the old one from the same tests. IMHO, the old error messages are MUCH more understandable (by the users who will see them as passed up via oiio) than the new ones. For example,

Old: `Failed OpenEXR read: Error reading pixel data from image file "src/incomplete.exr". Scan line 126 is missing.`

("Oh, a scanline is missing and I know which one. It's an incompletely written file.")

New: `EXR Error (src/incomplete.exr): EXR_ERR_BAD_CHUNK_DATA Request for scanline 0 found scanline 20000630, not 0 at chunk 0`

("What is a chunk and why does it think there is a scanline 20000630?")

**Comment by lgritz** [comment]
With the extra output reference checked in, CI is fully passing, so I'm going to merge.

Related items to-do:

* Benchmark to see if this has the performance gains (especially for texturing) that we hope for.
* Contemplate when we are comfortable using the core api automatically any time OpenEXR 3.1 is found.
* Should exrinput.cpp and exrinput_c.cpp be merged, with appropriate `#if`, rather than separate files? (Pro: will eliminate headache of edit in two places as other OIIO things change or we improve openexr handling. Con: possibly just as confusing to have to do it in two clauses of an `#if` and may look cluttered.)

**Pull request merged**
